service/history/hsm/tree.go

```diff
@@ -0,0 +1,366 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package hsm
+
+import (
+	"errors"
+	"fmt"
+
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+)
+
+// ErrStateMachineNotFound is returned when looking up a non-existing state machine in a [Node] or a [Collection].
+var ErrStateMachineNotFound = errors.New("state machine not found")
+
+// ErrStateMachineAlreadyExists is returned when trying to add a state machine with an ID that already exists in a [Collection].
+var ErrStateMachineAlreadyExists = errors.New("state machine already exists")
+
+// ErrIncompatibleType is returned when trying to cast a state machine's data to a type that it is incompatible with.
+var ErrIncompatibleType = errors.New("state machine data was cast into an incompatible type")
+
+// Key is used for looking up a state machine in a [Node].
+type Key struct {
+	// Type ID of the state machine.
+	Type int32
+	// ID of the state machine.
+	ID string
+}
+
+// State machine type.
+type MachineType struct {
+	// Type ID that is used to minimize the persistence storage space and address a machine (see also [Key]).
+	// Type IDs are expected to be immutable as they are used for looking up state machine definitions when loading data
+	// from persistence.
+	ID int32
+	// Human readable name for this type.
+	Name string
+}
+
+// StateMachineDefinition provides type information and a serializer for a state machine.
+type StateMachineDefinition interface {
+	Type() MachineType
+	// Serialize a state machine into bytes.
+	Serialize(any) ([]byte, error)
+	// Deserialize a state machine from bytes.
+	Deserialize([]byte) (any, error)
+}
+
+// cachedMachine contains deserialized data and state for a state machine in a [Node].
+type cachedMachine struct {
+	// An indicator that the data has not yet been loaded or has been marked stale and should be deserialized again.
+	dataLoaded bool
+	// Deserialized data.
+	data any
+	// Cached children.
+	children map[Key]*Node
+	outputs  []TransitionOutput
+}
+
+// Node is a node in a heirarchical state machine tree.
+//
+// It holds a persistent representation of itself and maintains an in-memory cache of deserialized data and child nodes.
+// Node data should not be manipulated directly and should only be done using [MachineTransition] or
+// [Collection.Transtion] to ensure the tree tracks dirty states and update transition counts.
+type Node struct {
+	// Key of this node in parent's map. Empty if node is the root.
+	Key Key
+	// Parent node. Nil if current node is the root.
+	Parent      *Node
+	registry    *Registry
+	cache       *cachedMachine
+	persistence *persistencespb.StateMachineNode
```

Comment:
Why are we keeping both the live and serialized version of a machine in memory at once? Is there really a benefit from serializing/deserializing them lazily at the cost of ~2x memory usage (probably more)?

Reply:
That's a good point.
I'm doing this so I can revert back to the old state when a transition mutates a state and fails.
A user may want handle that error and retry a different transition, leaving the state consistent.

Reply:
Yeah, I noticed that when I thought through how rollback works. The only reason our rollback works is because we discard the in-memory state; if we didn't we've nothing to prevent the access fn from both mutating the state _and_ returning an error, corrupting the state machine.

For rollback to work we need a pre-image to switch back to anyways, and we can either keep the serialized persistence version _or_ make a dynamic copy before running the transition function.